diff -rauB /hdd2/Dropbox/Olsr\ reactive/ns-2.34_olsr/olsr_modified/OLSR.cc ../olsr/OLSR.cc
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/olsr_modified/OLSR.cc	2013-12-16 22:52:20.000000000 +0000
+++ ../olsr/OLSR.cc	2013-12-18 00:17:03.497609971 +000
@@ -560,11 +560,15 @@
 	
 	// After processing all OLSR messages, we must recompute routing table
 	rtable_computation();
-	
+	post_rtable_computation(p);
 	// Release resources
 	Packet::free(p);
 }
 
+void
+OLSR::post_rtable_computation(Packet* p) {
+
+}
 ///
 /// \brief Computates MPR set of a node following RFC 3626 hints.
 ///
diff -rauB /hdd2/Dropbox/Olsr\ reactive/ns-2.34_olsr/olsr_modified/OLSR.h ../olsr/OLSR.h
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/olsr_modified/OLSR.h	2013-12-16 23:10:26.000000000 +0000
+++ ../olsr/OLSR.h	2013-12-18 00:14:47.877604939 +0000
@@ -312,7 +312,7 @@
 /// functionalities related to sending and receiving packets and managing
 /// internal state.
 ///
-class OLSR : public Agent {
+class OLSR : virtual public Agent {
 	// Makes some friends.
 	friend class OLSR_HelloTimer;
 	friend class OLSR_TcTimer;
@@ -324,6 +324,7 @@
 	friend class OLSR_TopologyTupleTimer;
 	friend class OLSR_IfaceAssocTupleTimer;
 	friend class OLSR_MsgTimer;
+	friend class Sally;
 	
 	/// Address of the routing agent.
 	nsaddr_t	ra_addr_;
@@ -393,13 +394,15 @@
 	
 	void		mpr_computation();
 	void		rtable_computation();
+	virtual void		post_rtable_computation(Packet*);
+
 	
 	void		process_hello(OLSR_msg&, nsaddr_t, nsaddr_t);
 	void		process_tc(OLSR_msg&, nsaddr_t);
 	void		process_mid(OLSR_msg&, nsaddr_t);
 	
 	void		forward_default(Packet*, OLSR_msg&, OLSR_dup_tuple*, nsaddr_t);
-	void		forward_data(Packet*);
+	virtual void		forward_data(Packet*);
 	
 	void		enque_msg(OLSR_msg&, double);
 	void		send_hello();
diff -drauB /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/aodv/aodv.h ../aodv/aodv.h
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/aodv/aodv.h	2009-06-14 18:35:07.000000000 +0100
+++ ../aodv/aodv.h	2013-12-17 21:21:57.109220030 +0000
@@ -191,7 +191,7 @@
 /*
   The Routing Agent
 */
-class AODV: public Agent {
+class AODV: virtual public Agent {
 
   /*
    * make some friends first 
@@ -216,7 +215,7 @@
         /*
          * Route Table Management
          */
-        void            rt_resolve(Packet *p);
+        virtual void            rt_resolve(Packet *p);
         void            rt_update(aodv_rt_entry *rt, u_int32_t seqnum,
 		     	  	u_int16_t metric, nsaddr_t nexthop,
 		      		double expire_time);
Only in ../aodv/: aodv_logs.o
Only in ../aodv/: aodv.o
Only in ../aodv/: aodv_rqueue.o
diff -drauB /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/aodv/aodv_rtable.h ../aodv/aodv_rtable.h
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/aodv/aodv_rtable.h	2009-06-14 18:35:07.000000000 +0100
+++ ../aodv/aodv_rtable.h	2013-12-18 01:04:49.685716353 +0000
@@ -81,6 +81,7 @@
 class aodv_rt_entry {
         friend class aodv_rtable;
         friend class AODV;
+        friend class Sally;
 	friend class LocalRepairTimer;
  public:
         aodv_rt_entry();
diff -drauB /hdd2/Dropbox/Olsr\ reactive/ns-2.34_olsr/Makefile.in ../Makefile.in
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/Makefile.in	2012-02-05 23:05:46.000000000 +0000
+++ ../Makefile.in	2013-12-16 16:39:45.548891631 +0000
@@ -277,6 +277,7 @@
 	aomdv/aomdv_logs.o aomdv/aomdv.o \
 	aomdv/aomdv_rtable.o aomdv/aomdv_rqueue.o \
 	olsr/OLSR.o olsr/OLSR_state.o olsr/OLSR_rtable.o olsr/OLSR_printer.o \
+    sally/sally.o \
 	common/ns-process.o \
 	satellite/satgeometry.o satellite/sathandoff.o \
 	satellite/satlink.o satellite/satnode.o \
diff -drauB /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-default.tcl ../tcl/lib/ns-default.tcl
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-default.tcl	2012-02-05 23:05:46.000000000 +0000
+++ ../tcl/lib/ns-default.tcl	2013-12-16 17:38:25.485022271 +0000
@@ -1371,6 +1371,14 @@
 Agent/OLSR set tc_ival_     5
 Agent/OLSR set mid_ival_    5
 
+# Defaults defined for UM-OLSR
+Agent/SALLY set debug_       false
+Agent/SALLY set use_mac_     false
+Agent/SALLY set willingness_ 3
+Agent/SALLY set hello_ival_  2
+Agent/SALLY set tc_ival_     5
+Agent/SALLY set mid_ival_    5
+
 # Intitialization for directed diffusion : Chalermek
 Agent/Diff_Sink set packetSize_ 512
 Agent/Diff_Sink set interval_   0.5
diff -drauB /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-lib.tcl ../tcl/lib/ns-lib.tcl
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-lib.tcl	2012-02-05 23:05:46.000000000 +0000
+++ ../tcl/lib/ns-lib.tcl	2013-12-16 17:26:44.748996263 +0000
@@ -636,6 +636,9 @@
 		    OLSR {
 			    set ragent [$self create-olsr-agent $node]
 		    }
+		    SALLY {
+			    set ragent [$self create-sally-agent $node]
+		    }
 		    TORA {
 			    Simulator set IMEPFlag_ ON
 			    set ragent [$self create-tora-agent $node]
@@ -870,6 +873,14 @@
 	$self at 0.0 "$ragent start"
 	$node set ragent_ $ragent
 	return $ragent
+}
+
+Simulator instproc create-sally-agent { node } {
+	#  Create UM-OLSR routing agent
+	set ragent [new Agent/SALLY [$node node-addr]]
+	$self at 0.0 "$ragent start"
+	$node set ragent_ $ragent
+	return $ragent
 }
 
 Simulator instproc use-newtrace {} {

diff -drauB /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-mobilenode.tcl ../tcl/lib/ns-mobilenode.tcl
--- /hdd2/Dropbox/Olsr reactive/ns-2.34_olsr/tcl/lib/ns-mobilenode.tcl	2009-06-14 18:35:41.000000000 +0100
+++ ../tcl/lib/ns-mobilenode.tcl	2013-12-17 16:04:04.460512152 +0000
@@ -198,6 +198,12 @@
 		$agent if-queue [$self set ifq_(0)]   ;# ifq between LL and MAC
 	}
 	
+	# Special processing for SALLY
+	set sallyonly [string first "SALLY" [$agent info class]] 
+	if {$sallyonly != -1 } {
+		$agent if-queue [$self set ifq_(0)]   ;# ifq between LL and MAC
+	}
+	
 	#<zheng: add>
 	# Special processing for ZBR
 	#set zbronly [string first "ZBR" [$agent info class]] 

